### ISSUE TITLE
Async function not type checking correctly

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#15830]
**TypeScript Version:**  2.3.2

**Code**

```ts
{
    const fn = (): Promise<number> => {
        const innerFn = async () => {
            if (1 === 1) {
                return Promise.resolve('wrong type');
            } else {
                return Promise.resolve(1);
            }
        };

        // should error with `Type 'string' is not assignable to type 'number'.`
        return innerFn();
    };

    fn();
}

```

**Expected behavior:**
`return innerFn()` should error with `Type 'string' is not assignable to type 'number'.`

**Actual behavior:**
`return innerFn()` does not error.

If you remove the `async` keyword from the async function, the program type checks as expected:

``` ts
{
    const fn = (): Promise<number> => {
        const innerFn = () => {
            if (1 === 1) {
                return Promise.resolve('wrong type');
            } else {
                return Promise.resolve(1);
            }
        };

        return innerFn();
    };

    fn();
}
```

**Comment by fis-cz** [comment]
Hi,

As your inner function does not specify the type I would say that default any rule is used.

If you really want to ensure the correct type to be returned it should be like:

```
{
    const fn = (): Promise<number> => {
        const innerFn = (): Promise<number> => {
            if (1 === 1) {
                return Promise.resolve('wrong type');
            } else {
                return Promise.resolve(1);
            }
        };

        return innerFn();
    };

    fn();
}
```

**Comment by OliverJAsh** [comment]
Hi @fis-cz. The inner type is inferred to have that return type, however the outer `fn` has a return type of `Promise<number>`, which `Promise<string|number>` is not assignable to.

**Comment by fis-cz** [comment]
when I do
```
{
    const fn = (): Promise<number> => {
        const innerFn = async (): Promise<number> => {
            if (1 === 1) {
                return Promise.resolve('wrong type');
            } else {
                return Promise.resolve(1);
            }
        };

        // should error with `Type 'string' is not assignable to type 'number'.`
        return innerFn();
    };

    fn();
}
```
it works correctly.

**Comment by OliverJAsh** [comment]
I don't always add type annotations for my function returns. I don't want it to silently fail when relying on inference.

**Comment by RyanCavanaugh** [comment]
Addressed by #15104

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
